Bugfixing kubeconfig decoding

janosdebugs opened the pull request:
## Changes introduced with this PR

This PR adds several bugfixes to the Kubeconfig parsing. Sadly, there has not been time to add tests for all edge cases yet.

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

Commit:
Bugfixing kubeconfig decoding

Commit:
Bugfixing kubeconfig decoding

Commit:
Bugfixing kubeconfig decoding

sandrobonazzola commented on kubeconfig_plugin.py:
```diff
@@ -111,66 +133,116 @@ class ErrorOutput:
     description=(
-        "Inputs a kubeconfig, parses it and extracts the kubernetes cluster"
-        " details "
+            "Inputs a kubeconfig, parses it and extracts the kubernetes cluster"
```
any reason for this additional indentation?

sandrobonazzola commented on tests/test_client_cert.yaml:
```diff
@@ -13,13 +13,13 @@ kubeconfig: |
-      client-key-data: LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUlFcEF
\ No newline at end of file
+      client-certificate-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUI0VENDQVl1Z0F3SUJBZ0lVQ0hoaGZmWTFsemV6R2F0WU1SMDJncEVKQ2hrd0RRWUpLb1pJaHZjTkFRRUwKQlFBd1JURUxNQWtHQTFVRUJoTUNRVlV4RXpBUkJnTlZCQWdNQ2xOdmJXVXRVM1JoZEdVeElUQWZCZ05WQkFvTQpHRWx1ZEdWeWJtVjBJRmRwWkdkcGRITWdVSFI1SUV4MFpEQWVGdzB5TWpBNU1qZ3dOVEk0TVRKYUZ3MHlNekE1Ck1qZ3dOVEk0TVRKYU1FVXhDekFKQmdOVkJBWVRBa0ZWTVJNd0VRWURWUVFJREFwVGIyMWxM ...
+      client-key-data: LS0tLS1CRUdJTiBQUklWQVRFIEtFWS0tLS0tCk1JSUJWQUlCQURBTkJna3Foa2lHOXcwQkFRRUZBQVNDQVQ0d2dnRTZBZ0VBQWtFQXJyODlmMmtnZ1NPL3lhQ0IKNkV3SVFlVDZacHRCb1gwWnZDTUkrRHBrQ3dxT1M1ZndSYmoxbkVpUG5MYnpERGdNVThLQ1BBTWhJN0pwWVJsSApuaXB4V3dJREFRQUJBa0J5YnUveDBNRWxjR2kydS9KMlVkd1Njc1Y3amU1VHQxMno4Mmw3VEptWkZGSjhSTG1jCnJoMDBHdmViNFZwR2hkMStjM2xaYk8xbUlUNnYzdkhNOUEwaEFpRUExNEVXNmIrOTlYWXphNys1dXdJ ...
```
missing new line at end of file

sandrobonazzola reviewed (changes requested):
A few inline comments. Also https://github.com/arcalot/arcaflow-plugin-kubeconfig/actions/runs/3272042293/jobs/5382567090#step:4:27

engelmi commented on kubeconfig_plugin.py:
```diff
@@ -117,57 +138,106 @@ class ErrorOutput:
+        try:
+            context = None
+            for ctx in kubeconfig["contexts"]:
+                if ctx["name"] == current_context:
```
Just to make sure: `current_context` from `kubeconfig["current-context"]` is a `string` and not a `dict`, right?

Harshith-umesh commented on kubeconfig_plugin.py:
```diff
@@ -117,57 +138,106 @@ class ErrorOutput:
-    except Exception:
-        return "error", ErrorOutput(1, "Failure in parsing kubeconfig:")
+        output = SuccessOutput(
+            Connection(cluster["server"]),
```
can we first assign cluster["server"] as host and then use that, it'll reduce confusion in the future

engelmi commented on kubeconfig_plugin.py:
```diff
@@ -117,57 +138,106 @@ class ErrorOutput:
+                                                        ]).decode('ascii')
+        except KeyError:
+            pass
+        try:
+            output.connection.cert = base64.b64decode(user[
+                                                          "client-certificate-data"
+                                                      ]).decode('ascii')
+        except KeyError:
+            pass
```
Could you wrap thos in a dedicated function? 
```python
def getFromDict(d -> dict, key -> str) -> str:
   try:
      return d[key]
   except KeyError:
      pass
```

portante replied:
Why can't one use:
```
d.get(key, "")
```

portante replied:
If you can't have empty fields in the `Connection` data object then you need the try/except style.

jaredoconnell replied:
The format that the schema needs is `None` for all missing optional values. We can't use get for this, so the only ways for this to work is the current way, or a just-as-verbose `in` conditional.

jaredoconnell replied:
Fixed in next commit (in progress).

dustinblack reviewed (approved):
I only did a cursory code review due to time constraints, but I did do a functional test, and my workflow was able to read the kubeconfig and then deploy a pod into a kubernetes cluster.

dustinblack commented on requirements.txt:
```diff
@@ -1,28 +1,3 @@
 arcaflow-plugin-sdk==0.8.0; python_version >= "3.9"
```
Bump this to 0.9.0

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
-        "Inputs a kubeconfig, parses it and extracts the kubernetes cluster"
-        " details "
+            "Inputs a kubeconfig, parses it and extracts the kubernetes cluster"
+            " details "
```
`s/ details / details/`

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
+            current_context = kubeconfig["current-context"]
+        except KeyError:
+            return "error", ErrorOutput(
+                "The provided kubeconfig file does not have a current-context set. Please set a current context to use.")
```
Consider the following to reduce the error returns and lines of code a bit:

```
kind = kubeconfig.get("kind", "")
if kind != "Config":
    addendum = "" if kind else " (missing 'kind' field)"
    return "error", ErrorOutput(f"The provided file is not a kubeconfig file{addendum}")
current_context = kubeconfig.get("current-context", "")
if not current_context:
    return "error", ErrorOutput(
                "The provided kubeconfig file does not have a current-context set. Please set a current context to use.")
```

jaredoconnell replied:
Code refactored.

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
+                    "Failed to find a context named {} in the kubeconfig file.".format(current_context))
+            current_cluster = context["cluster"]
+            current_user = context["user"]
+        except Exception as e:
```
Shouldn't this be `except KeyError as e`?

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
             return "error", ErrorOutput(
-                1, "Both client data and token missing, exiting!"
-            )
+                "Failed to find a context named {} in the kubeconfig file: {}".format(current_context, e.__str__()))
```
Consider:

```
return "error", ErrorOutput(f"Failed to find a context named {current_context} in the kubeconfig file: {e!r}")
```

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
+                    "Failed to find a cluster named {} in the kubeconfig file.".format(current_cluster))
+        except Exception as e:
+            return "error", ErrorOutput(
+                "Failed to find a cluster named {} in the kubeconfig file: {}".format(current_cluster, e.__str__()))
```
Consider a bug fix (need to prevent `KeyError` in the inner loop):

```
cluster = None
error_addendum = ""
try:
    clusters = kubeconfig["clusters"]
except Exception as e:
    error_addendum = f": {e}"
else:
    for cl in clusters:
        try:
            if cl["name"] == current_cluster:
                cluster = cl["cluster"]
        except KeyError:
            pass
if cluster is None:
    return "error", ErrorOutput(
        f"Failed to find a cluster named {current_cluster} in the kubeconfig file{error_addendum}"
    )
```

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
+                    "Failed to find a user named {} in the kubeconfig file.".format(current_user))
+        except Exception as e:
+            return "error", ErrorOutput(
+                "Failed to find a user named {} in the kubeconfig file: {}".format(current_user, e.__str__()))
```
Consider a similar refactoring of lines 163 - 173 to what is proposed for changes in the comment for line 161.

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +109,107 @@ class ErrorOutput:
-    except Exception:
-        return "error", ErrorOutput(1, "Failure in parsing kubeconfig:")
+        output = SuccessOutput(
+            Connection(host=cluster["server"]),
```
If `server` is not found in, shouldn't this report the `KeyError` without a huge stack that you'd get from the exception handling at line 210?  And perhaps consider moving this up to immediately follow lines 161?

jaredoconnell replied:
Now handled with explicit check above.

portante reviewed (changes requested):
One potential bug, consider some alternative approaches in the code.

portante reviewed (commented):
Latest changes look fine.

jaredoconnell committed:
Address some PR comments

portante reviewed (commented):
Looking good.

dustinblack commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +109,92 @@ class ErrorOutput:
+        output = SuccessOutput(
+            Connection(host=cluster["server"]),
+        )
+	output.connection.cacert = base64_decode(cluster.get("certificate-authority-data", None))
```
Leading white space here is a tab, not spaces, throwing an error when `test_kubeconfig_plugin.py` is run.

jaredoconnell committed:
Fix tab in space-indented file

jaredoconnell committed:
Fix bug with base64_decode, and use dev SDK

jaredoconnell committed:
Improve formatting, and some messages

Added flake8 file to align with the black formatter

portante commented on pyproject.toml:
```diff
@@ -2,16 +2,16 @@
 [tool.poetry.dependencies]
 python = "^3.9"
-arcaflow-plugin-sdk = "0.9.0"
+arcaflow-plugin-sdk = { git = "https://github.com/arcalot/arcaflow-plugin-sdk-python.git", branch="allow-scope-in-one-of" }
```
Why are we not first merging the `allow-scope-in-one-of` branch to `main` in the `arcaflow-plugin-sdk-python` respository first?

dustinblack replied:
Clearly we don't want to merge this PR until we are ready to replace this with a version number. We should convert this to a draft for now.

portante reviewed (commented):
Any consideration for the existing comments?

dustinblack committed:
enable carpenter builds for development branches

jaredoconnell committed:
Added test cases to ensure that code works properly

These would have caught the errors introduced in a prior commit

jaredoconnell committed:
Fix linter issues

jaredoconnell committed:
Refactored code to have explicit key checks

Also added tests for each of the cases, and general tests

Also updated poetry lock

jaredoconnell commented:
I refactored the code and added a lot of test cases. The test cases cover the general cases, as well as each of the potential missing key situations.

dustinblack committed:
Merge pull request #6 from arcalot/branch-builder

enable carpenter builds for development branches

jaredoconnell committed:
Added gitleaks allow file

This is to prevent false positives with the info-sec service

jaredoconnell committed:
Revert to latest release of SDK

jaredoconnell committed:
Merge remote-tracking branch 'origin/bugfix' into bugfix

jaredoconnell committed:
Updated poetry lock

jaredoconnell reviewed (approved):
Confirmed to work with old SDK. I believe this is ready for merging. The extra checks make this a much safer option than the code in main.

sandrobonazzola reviewed (approved):
LGTM

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
+        except Exception as e:
+            return "error", ErrorOutput(
+                "Exception occurred while loading YAML. Input is not valid YAML."
+                + " Exception: {}".format(e.__str__())
```
Wouldn't an f-string work here?

```
            return "error", ErrorOutput(
                "Exception occurred while loading YAML. Input is not valid YAML."
                f" Exception: {e}"
            )
```

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
+        else:
+            return "error", ErrorOutput(
+                "The provided file is not a kubeconfig file (missing 'kind' field)"
+            )
```
Consider a couple of alternatives if you are gonna be back in here:

```
        if "kind" not in kubeconfig:
            return "error", ErrorOutput(
                "The provided file is not a kubeconfig file (missing 'kind' field)"
            )
        if kubeconfig["kind"] != "Config":
            return "error", ErrorOutput(
                "The provided file is not a kubeconfig file"
            )
```

or a bit longer but you leverage a more pythonic way to do this where you only reference the dictionary once:

```
        try:
            kind = kubeconfig["kind"]
        except KeyError:
            return "error", ErrorOutput(
                "The provided file is not a kubeconfig file (missing 'kind' field)"
            )
        if kind != "Config":
            return "error", ErrorOutput(
                "The provided file is not a kubeconfig file"
            )
```

portante replied:
And this notion of checking a dictionary and then accessing the dictionary for the same element is littered below.  It is usually worth the effort to only access the dictionary once.

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
+        current_context = kubeconfig.get("current-context", None)
+        if current_context is None:
+            return "error", ErrorOutput(
+                """The provided kubeconfig file does not have a current-context set."""
```
What about these strings requires the triple quotes?

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
             return "error", ErrorOutput(
-                1, "Both client data and token missing, exiting!"
+                "Failed to find a context named {} in the kubeconfig file.".format(
+                    current_context
```
Is there a reason you can't use an f-string here?

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
+        for expected_key in ["cluster", "user"]:
+            if expected_key not in context:
+                return "error", ErrorOutput(
+                    "'{}' field missing from kubeconfig current context.".format(
```
Perhaps an f-string?

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
+                if "cluster" not in cl:
+                    return "error", ErrorOutput(
+                        "cluster section missing from section "
+                        + "of current cluster {}".format(current_cluster)
```
Python allows you to place string "constants" next to each other to auto combine them.  So something like:

```
"this and" " that"
```

Is equivalent to `"this and that"`.  This also applies across line boundaries, and between f-strings and non-f-strings.  So you can drop the plus sign do just do:

```
"cluster section missing from section of"
f" current cluster {current_cluster}"
```

jaredoconnell replied:
Thanks for the Python tips!

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
+                cluster = cl["cluster"]
+        if cluster is None:
+            return "error", ErrorOutput(
+                "Failed to find a cluster named {} in the kubeconfig file.".format(
```
f-string?

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
+            for required_key in ["name", "user"]:
+                if required_key not in u:
+                    return "error", ErrorOutput(
+                        """'{0}' section in the users section not found in the """
```
What requires the triple quotes?

And can we use an f-string here?

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
+                user = u["user"]
+        if user is None:
+            return "error", ErrorOutput(
+                "Failed to find a user named {} in the kubeconfig file.".format(
```
f-string?

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
+        # The goal is for all other errors to be addressed individually.
+        print(traceback.format_exc())
+        return "error", ErrorOutput(
+            "Failure to parse kubeconfig. Exception: {}".format(e.__str__())
```
f-string?

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
+        if "server" not in cluster:
+            return "error", ErrorOutput(
+                """Failed to find server in cluster kubeconfig file {0}"""
+                """ cluster section""".format(current_cluster)
```
Same comment as above ...

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
+    except Exception as e:
+        # This is the catch-all case.
+        # The goal is for all other errors to be addressed individually.
+        print(traceback.format_exc())
```
Why are we printing a traceback and returning an error?

If we do need to print an error, shouldn't it be on `stderr`?

Wouldn't we want to format the stack trace into the exception string returned?  Will the person using the plug-in be able to find the output printed to stdout separate from that returned by the plug-in?

jaredoconnell replied:
Added to the error output.

portante commented on kubeconfig_plugin.py:
```diff
@@ -111,63 +111,151 @@ class ErrorOutput:
+                    )
+                )
+        current_cluster = context["cluster"]
+        current_user = context["user"]
```
Instead of a whole for-loop, could we consider something a bit simpler which will make it easier to understand:

```
        try:
            current_cluster = context["cluster"]
            current_user = context["user"]
        except KeyError as e:
            return "error", ErrorOutput(
                f"{e} field missing from kubeconfig current context."
            )
```

portante reviewed (dismissed):
In combination it seems it's worth to consider these suggested changes for maintainability of the code.

jaredoconnell committed:
Use better Python practices
